### PR TITLE
Aws sdk v2

### PR DESCRIPTION
**Pull request by danielwanja**
* Added s3 storage implementation using the aws-sdk v2.

### CODE REVIEW

File: lib/paperclip/storage/s3v2.rb [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+            end
+
+            write_options[:metadata] = @s3_metadata unless @s3_metadata.empty?
+            write_options.merge!(@s3_headers)
+
+            s3_object(style).upload_file(file.path, write_options)
+          rescue Aws::S3::Errors::NoSuchBucket
+            create_bucket
+            retry
+          rescue Aws::S3::Errors::SlowDown    # Daniel: SlowDown not defined in V2. See what concept replaces it. There is a Waiters concept
```
Comment (danielwanja) [inline lib/paperclip/storage/s3v2.rb]: Actually the Watchers are a new concept but don't seem to match the us of the SlowDown exception and I din't think we can use them to the same end. My thought was to just remove the SlowDown exception rescuing as it's not thrown by the V2 api anymore.  Note the V2 SDK has now a built in retry_limit (3 by default).

**Comment by booleanbetrayal** [comment]
hope this one gets through sometime soon!  =]

**Comment by danielwanja** [comment]
I'm still working on it. Last week I adapted to the official release of the
awl SDK v2. Fixed a usage issue when used with delayed_paperclip. Didn't
push if yet, still having issues with the s3_region specs. We are using
this branch in prod at blinker.com.

On Friday, February 27, 2015, Brent <notifications@github.com> wrote:

> hope this one gets through sometime soon! =]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thoughtbot/paperclip/pull/1730#issuecomment-76398829>.
>


-- 
Sent from my iPhone

File: lib/paperclip/storage/s3v2.rb [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+      def self.extended base
+        begin
+          require 'aws-sdk'
+        rescue LoadError => e
+          e.message << " (You may need to install the aws-sdk gem)"
+          raise e
+        end unless defined?(Aws::S3)
+
+        # Overriding log formatter to make sure it return a UTF-8 string
+        if defined?(Seahorse::Client::Logging::Formatter)
```
Comment (danielwanja) [inline lib/paperclip/storage/s3v2.rb]: This is in line with the code in s3.rb (https://github.com/thoughtbot/paperclip/blob/master/lib/paperclip/storage/s3.rb#L123). Just using their new formatter.

**Comment by maclover7** [comment]
Nice PR! Can you start out by fixing the Hound style guide issues?

**Comment by danielwanja** [comment]
Hi Jon, I will work on these.  I assume you added hound later in the game as I get a bunch of errors against the original s3.rb code on which this is based. Also I still have one spec that fails on travis-ci but of course works locally, so I'll try to reproduce that one too.

**Comment by danielwanja** [comment]
Do you guys really want me to fix all the houdci notifications. Even the line too long ones? Make code less readable in my view? What else would you like to see fixed for this to be accepted?

**Comment by booleanbetrayal** [comment]
some of those houndci warnings are super aggressive. would be nice to see this changeset get in!

**Comment by maclover7** [comment]
@tute @jyurek This should probably come in 5.0, right?

**Comment by tute** [comment]
@danielwanja thank you for your work!
@jyurek what do you think about this changeset?

**Comment by danielwanja** [comment]
@tute, my pleasure. Let me know what you thing of this changeset and what would be needed to have this merged.

**Comment by nathantechie9** [comment]
I would like to help with this. I already mocked up a solution for the was s3 gem at https://github.com/marcel/aws-s3/pull/108

**Comment by jyurek** [comment]
@nathantechie9 There is no change that needs to happen with the aws-s3 gem. Paperclip has been using the aws-sdk gem for some time now, which is a different gem, but aside from that, the library should not need to change to accommodate us.

**Comment by jyurek** [comment]
@danielwanja I would really appreciate it if you could look at the places Hound says are too complicated and refactor them a bit. Thanks! :)

**Comment by nathantechie9** [comment]
What's the issue with the regions? I'd like to help.

**Comment by tute** [comment]
Is the functionality in this PR comparable to that of https://github.com/thoughtbot/paperclip/pull/1903?

**Comment by danielwanja** [comment]
Same idea, to use the new aws-sdk v2, different approach. PR #1903
<https://github.com/thoughtbot/paperclip/pull/1903> intermingles V1 and V2
code in the same adapter using checking *aws_v1?* and calling the
appropriate underlying aws-sdk method. I preferred creating a new adapter
that doesn't mingle both. Being originally from Switzerland I can see
advantages in both approaches.  The main goal for me was to be able to use
some of the new features the aws-sdk v2 prodives, notably using the  new
kms related headers ("x-amz-server-side-encryption" => "aws:kms",
"x-amz-ssekms-key-id"
=> "alias/mycustom_encryption_key").

**Comment by tute** [comment]
AWS 2 support has been merged: https://github.com/thoughtbot/paperclip/pull/1903#issuecomment-133160925. Thanks for your work!

**Pull request closed without merging**